File: tyto/go/core/typeutil/int.go
package typeutil

// 用high和low组成一个新的int，high放高位，low放低位
func MakeInt16(high int8, low int8) int16 {
	// int8(-1) 和 int16(-1) 的二进制数据是不同的
	// 因此，先转换为同长度的无符号类型
	data := uint16(uint8(high)) << 8
	data |= uint16(uint8(low))
	return int16(data)
}

func MakeInt32(high int16, low int16) int32 {
	data := uint32(uint16(high)) << 16
	data |= uint32(uint16(low))
	return int32(data)
}

func MakeInt64(high int32, low int32) int64 {
	data := uint64(uint32(high)) << 32
	data |= uint64(uint32(low))
	return int64(data)
}

// 获取高位部分
func HighPartInt16(i int16) int8 {
	data := uint16(i) >> 8
	return int8(data)
}

func HighPartInt32(i int32) int16 {
	data := uint32(i) >> 16
	return int16(data)
}

func HighPartInt64(i int64) int32 {
	data := uint64(i) >> 32
	return int32(data)
}

// 获取低位部分
func LowPartInt16(i int16) int8 {
	data := uint16(i) & 0xff
	return int8(data)
}

func LowPartInt32(i int32) int16 {
	data := uint32(i) & 0xffff
	return int16(data)
}

func LowPartInt64(i int64) int32 {
	data := uint64(i) & 0xffffffff
	return int32(data)
}

File: tyto/csharp/core/Tyto.Utils/TypeUtil/Int.cs
namespace Tyto.Utils;

using int8 = sbyte;
using int16 = short;
using int32 = int;
using int64 = long;
using uint8 = byte;
using uint16 = ushort;
using uint32 = uint;
using uint64 = ulong;

/// <summary>
/// 整数合并和分拆处理类
/// </summary>
public static partial class TypeUtil
{
	/// <summary>
	/// 将high和low组合为一个新的整数，high放高位部分，low放低位部分
	/// </summary>
	public static int16 MakeInt16(int8 high, int8 low)
	{
		uint16 h = (uint8)high;
		uint16 l = (uint8)low;
		return (int16)((h << 8) | l);
	}

	/// <inheritdoc cref="MakeInt16"/>
	public static int32 MakeInt32(int16 high, int16 low)
	{
		uint32 h = (uint16)high;
		uint32 l = (uint16)low;
		return (int32)((h << 16) | l);
	}

	/// <inheritdoc cref="MakeInt16"/>
	public static int64 MakeInt64(int32 high, int32 low)
	{
		uint64 h = (uint32)high;
		uint64 l = (uint32)low;
		return (int64)((h << 32) | l);
	}

	/// <summary>
	/// 获取整数的高位部分
	/// </summary>
	public static int8 HighPartInt16(int16 val)
	{
		return (int8)((uint16)val >> 8);
	}

	/// <inheritdoc cref="HighPartInt16"/>
	public static int16 HighPartInt32(int32 val)
	{
		return (int16)((uint32)val >> 16);
	}

	/// <inheritdoc cref="HighPartInt16"/>
	public static int32 HighPartInt64(int64 val)
	{
		return (int32)((uint64)val >> 32);
	}

	/// <summary>
	/// 获取整数的低位部分
	/// </summary>
	public static int8 LowPartInt16(int16 val)
	{
		return (int8)((uint16)val & 0xFF);
	}

	/// <inheritdoc cref="LowPartInt16"/>
	public static int16 LowPartInt32(int32 val)
	{
		return (int16)((uint32)val & 0xFFFF);
	}

	/// <inheritdoc cref="LowPartInt16"/>
	public static int32 LowPartInt64(int64 val)
	{
		return (int32)((uint64)val & 0xFFFFffff);
	}
}
